File: php/path_traversal_vulnerable.php
<?php
// Path Traversal Vulnerable Test Cases

// Matches: php-path-traversal-file-get-contents
function vulnerable_file_get_contents($userInput) {
    return file_get_contents($userInput);
}

// Matches: php-path-traversal-fopen
function vulnerable_fopen($userInput) {
    return fopen($userInput, "r");
}

// Matches: php-path-traversal-readfile
function vulnerable_readfile($userInput) {
    readfile($userInput);
}

// Matches: php-path-traversal-include
function vulnerable_include($userInput) {
    include($userInput);
}

// Safe: basename and allowlist
function safe_read($userInput) {
    $allowed = ['page1.php', 'page2.php'];
    $file = basename($userInput);
    if (in_array($file, $allowed)) {
        return file_get_contents("/safe/dir/" . $file);
    }
    return false;
}

File: go/path_traversal_vulnerable.go
package main

import (
	"os"
	"path/filepath"
)

// Matches: go-path-traversal-os-open
func vulnerableOsOpen(userInput string) (*os.File, error) {
	return os.Open(userInput)
}

// Matches: go-path-traversal-readfile
func vulnerableReadFile(userInput string) ([]byte, error) {
	return os.ReadFile(userInput)
}

// Matches: go-path-traversal-filepath-join
func vulnerableFilepathJoin(userInput string) string {
	return filepath.Join("/uploads", userInput)
}

// Safe: validate and clean
func safeRead(userInput string) ([]byte, error) {
	clean := filepath.Clean(userInput)
	abs := filepath.Join("/safe/dir", clean)
	if !filepath.HasPrefix(abs, "/safe/dir") {
		return nil, fmt.Errorf("path traversal detected")
	}
	return os.ReadFile(abs)
}
